mne/stats/cluster_level.py

```diff
@@ -1729,3 +1741,363 @@ def summarize_clusters_stc(
     data_summary[:, 0] = np.sum(data_summary, axis=1)
 
     return klass(data_summary, vertices, tmin, tstep, subject)
+
+
+def _validate_cluster_df(df: pd.DataFrame, dv_name: str, iv_name: str):
+    """Validate the input DataFrame for cluster tests."""
+    # check if all necessary columns are present
+    missing = ({dv_name} | {iv_name}) - set(df.columns)  # should be empty
+    sep = '", "'
+    if missing:  # if not empty, there are missing columns
+        raise ValueError(
+            f"DataFrame must contain a column named for each term in `formula`. "
+            f"Column{_pl(missing)} missing for term{_pl(missing)} "  # _pl = pluralize
+            f'"{sep.join(missing)}".'
+        )
+    # check if the data column contains valid (and consistent) instance types
+    inst = df[dv_name].iloc[0]
+    valid_types = (
+        Evoked,
+        BaseEpochs,
+        BaseTFR,
+        np.ndarray,
+    )  # Base covers all Epochs and TFRs
+    _validate_type(inst, valid_types, f"Data in dependent variable column '{dv_name}'")
+    all_types = set(df[dv_name].map(type))
+    all_type_names = ", ".join([type(x).__name__ for x in all_types])
+    prologue = f"Data in dependent variable column '{dv_name}' must all have "
+    if len(all_types) > 1:
+        raise ValueError(
+            f"{prologue} the same type, but found types {{{all_type_names}}}."
+        )
+    # check if the shape of the data is consistent
+    if isinstance(inst, np.ndarray):
+        all_shapes = set(
+            df[dv_name].map(lambda x: x.shape[1:])
+        )  # first dim may vary (participants or epochs)
+    elif isinstance(inst, (BaseEpochs | BaseTFR)):
+        all_shapes = set(df[dv_name].map(lambda x: x.get_data().shape[1:]))
+    else:
+        all_shapes = set(df[dv_name].map(lambda x: x.get_data().shape))
+    if len(all_shapes) > 1:
+        raise ValueError(
+            f"{prologue} consistent shape, but {len(all_shapes)} different "
+            f"shapes were found: {'; '.join(all_shapes)}."
+        )
+    return all_types.pop()  # return the type of the data column entries
+
+
+@verbose
+def cluster_test(
+    df: pd.DataFrame,
+    formula: str,
+    *,  # end of positional-only parameters
+    within_id: str | None = None,
+    stat_fun: callable | None = None,
+    tail: Literal[-1, 0, 1] = 0,
+    threshold=None,
+    n_permutations: str | int = 1024,
+    adjacency: sparse.spmatrix | None | False = None,  # should be None (default)
```

Comment:
If only `False`, but not any bool is accepted, should be typed as:
```suggestion
    adjacency: sparse.spmatrix | None | Literal[False] = None,  # should be None (default)
```